add Logs plugin to view Sia logs in UI

avahowell opened the pull request:
This PR implements #449, and the currently open project on this repo. It adds a `Logs` plugin, which can be used to view each siad log, or combinations of the logs, defined by `filters.js`.

Screenshots:
![sia1](https://cloud.githubusercontent.com/assets/8183920/20230130/56187626-a828-11e6-8a54-cf7fe2acd7fe.png)
![sia2](https://cloud.githubusercontent.com/assets/8183920/20230131/5618cb44-a828-11e6-936d-9e218f70c66f.png)
![sia3](https://cloud.githubusercontent.com/assets/8183920/20230129/56158948-a828-11e6-9ccb-419671c7fbef.png)
![sia4](https://cloud.githubusercontent.com/assets/8183920/20230128/561570c0-a828-11e6-8b3a-5e2bb7e05abf.png)

avahowell committed:
initial commit, add Logs plugin

avahowell committed:
add filtercontrol component, statically define filter control buttons

avahowell committed:
default log text filters, use reduce in findLogs

avahowell committed:
add inline styles, move helpers to logparse.js

avahowell committed:
move onFilterClick to filterControl

avahowell committed:
enable multiple filtercontrol selection

avahowell committed:
order logs plugin in plugin list

avahowell committed:
final styling tweaks

avahowell committed:
move filterButtons to ./filters.js

avahowell committed:
fix `filtered` shadowing in FilterControl

avahowell committed:
add siad output log filter

lukechampine commented on js/rendererjs/plugins.js:
```diff
@@ -116,35 +116,31 @@ export const scanFolder = (path) => {
-		}
-		return 0
-	})
+	plugins = pushToTop(plugins, homePlugin)
```
this is an interesting way of doing things, haha.
Maybe instead we could map each plugin to a "priority" number?
The problem there, I suppose, is that you want an easy way to stick plugins at _either end_ of the list, not just the top (or just the bottom). So a priority list isn't a perfect fit because it tends to only have one "maximum" value (e.g. 0 means highest priority).

lukechampine commented on plugins/Logs/js/components/filtercontrols.js:
```diff
@@ -0,0 +1,42 @@
+				<FilterControl
+					name={filter.name}
+					filters={filter.filters}
+					checked={filter.filters.reduce((isChecked, filtertext) => isChecked || logFilters.includes(filtertext), false)}
```
congrats, "filter" is now semantically satiated.

lukechampine commented on plugins/Logs/js/components/filtercontrol.js:
```diff
@@ -0,0 +1,43 @@
+	}
+	return (
+		<div style={filterControlStyle} onClick={onFilterClick(filters)}>
+			<span style={{WebkitUserSelect: 'none'}}> {name} </span>
```
what does this style do?

avahowell replied:
It prevents the button names from being selected when you shift-click.

lukechampine commented on plugins/Logs/js/components/logview.js:
```diff
@@ -0,0 +1,28 @@
+const LogView = ({logText}) => (
+	<div style={logViewStyle}>
+		{logText}
+	</div>
```
too complicated

lukechampine commented:
My main concerns are:
- When does the log output get refreshed?
- Is it dangerous to read the entire log file? What if the file is 10MB?

I think ideally the log files would be treated as streams, rather than reading them all into memory. Also, it would be nice if it were possible to only show the output from the current session (maybe via a toggle). Otherwise if you have a long-running daemon you can wind up reading tons of data that you don't care about. Not a pressing concern though.

DavidVorick commented:
Can we move the bar to the top?

DavidVorick commented:
LOGS: m1 | m2 | etc

DavidVorick commented:
DEBUG logs can hit hundreds of MBs

avahowell committed:
add readLog to seek and read partial log files, logging testing

avahowell committed:
add cleanLog to remove partial log lines

avahowell committed:
change readLog to take nbytes

avahowell committed:
readLog now reads the last `nbytes`, logView autoscrolls

avahowell committed:
increment log size if scroll reaches top, poll for log updates

avahowell commented:
As of latest commit:
- Only the last `n` bytes of logs are loaded into memory. `n` defaults at 50KB.
- If you hit the top of the log view scrolling, 50 more KB of logs will be loaded.

avahowell committed:
update electron-prebuilt -> electron in test/app.js

avahowell committed:
fix readLog len undefined case

lukechampine commented on plugins/Logs/js/components/filtercontrol.js:
```diff
@@ -0,0 +1,43 @@
+			removeLogFilters(clickedFilters)
+			return
+		}
+		addLogFilters(clickedFilters)
```
I think making each filter toggle would be more natural. If I see a row of buttons, my UX-sense doesn't expect that I'll be able to shift-click them.

lukechampine commented on plugins/Logs/js/components/logview.js:
```diff
@@ -0,0 +1,51 @@
+	componentDidMount() {
+		this._logView.scrollTop = this._logView.scrollHeight
+		this._logView.addEventListener('scroll', this.handleScroll.bind(this))
+	}
```
tragic

lukechampine commented on plugins/Logs/js/logparse.js:
```diff
@@ -0,0 +1,48 @@
+const siadir = SiaAPI.config.siad.datadir
+
+const fileSize = (logpath) =>
+	fs.statSync(logpath).size
```
I usually stick short arrow function bodies on the same line if they're just a single expression. It's arguably less readable, but with a newline+indent it looks like python :P

avahowell replied:
yeah, in this case I think a single line is prettier.

lukechampine commented on plugins/Logs/js/logparse.js:
```diff
@@ -0,0 +1,48 @@
+	const fd = fs.openSync(logpath, 'r')
+	fs.readSync(fd, buf, 0, len, startPos)
+
+	return cleanLog(buf.toString())
```
might be better to slice the buffer instead of the string. I doubt there's much performance impact though

lukechampine replied:
`readSync` can throw an exception, right? Is it caught anywhere?

avahowell committed:
set the upper bound of len to min(nbytes, filesize)

avahowell committed:
add logs plugin integration tests

avahowell committed:
catch exceptions from readSync

avahowell committed:
single-line cleanLog and fileSize arrow functions

avahowell committed:
change filter controls to toggle by default

avahowell commented:
As of the latest commit, I've added more comprehensive testing, handled the exception from `readSync`, and changed the default button behavior to be a toggle.

avahowell committed:
move filter controls to top

avahowell committed:
change toggle behavior back to ctrl select

avahowell committed:
hide Logs plugin by default

avahowell committed:
remove Logs plugin sidebar ordering

avahowell committed:
only remove log filter if there is more than 1 selected

avahowell commented:
Logs plugin is now hidden from the sidebar and can be activated by pressing `Control + Shift + L`. After some discussion and testing, the previous behavior of exclusively selecting logs on click and toggling them on ctrl-click or shift-click has been restored, and the buttons have been moved to the top.
